File: im-ai-office/imai/services/digest.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""每日汇总兜底推送（M2 收尾；Spec：每日汇总兜底Spec.md）

产品承诺（一页纸 M2）：发送者当天未确认 → 下班前自动把待确认清单推给群主/管理员。
此前仅有 /api/summary/daily 被动接口（生成不推送），本模块补齐闭环：

- 调度：scheduler 每轮调用 scan_and_push；本地时间 ≥ DIGEST_TIME（默认 18:00）
  且当日未推过 → 推送一次。digest_sent 按日期主键幂等，重启/重复扫描不重发。
- 收件人：role 表 role='admin'（RBAC 管理员）；为空时回落 DIGEST_FALLBACK_ADMIN。
- 通道：ai_dm（AI 助手会话，UI 可见）+ SSE fanout + audit 留痕，与到期提醒同型；
  不打群（防骚扰原则）。
- 时间语义：now 为本地 naive datetime（与 reminder.scan_once 同约定）。
"""
from datetime import datetime

from imai.config import DIGEST_TIME, DIGEST_FALLBACK_ADMIN
from imai.repos import audit_log
from imai.services.ai_dm import ai_dm_send
from imai.services import bus
from imai.services.memory import build_daily_summary


def _digest_admins(con):
    """RBAC 管理员收件人；role 表无 admin 时回落配置兜底人。"""
    c = con.cursor()
    try:
        c.execute("SELECT oim_user_id FROM role WHERE role='admin' ORDER BY oim_user_id")
        ids = [r[0] for r in c.fetchall()]
    except Exception:
        ids = []
    return [i for i in ids if i] or [DIGEST_FALLBACK_ADMIN]


def _already_pushed(con, date_str):
    c = con.cursor()
    c.execute("SELECT 1 FROM digest_sent WHERE digest_date=?", (date_str,))
    return c.fetchone() is not None


def scan_and_push(con, now=None):
    """每轮调度调用：到点且当日未推 → 生成汇总并推给管理员。

    返回 {"pushed": bool, "date": str, ...}；未推时附 reason（before_time/already_sent）。
    """
    now = now or datetime.now()
    date_str = now.strftime("%Y-%m-%d")
    if _already_pushed(con, date_str):
        return {"pushed": False, "reason": "already_sent", "date": date_str}
    try:
        hh, mm = DIGEST_TIME.split(":")[:2]
        gate = now.replace(hour=int(hh), minute=int(mm), second=0, microsecond=0)
    except (ValueError, IndexError):
        gate = now.replace(hour=18, minute=0, second=0, microsecond=0)
    if now < gate:
        return {"pushed": False, "reason": "before_time", "date": date_str}

    sm = build_daily_summary(con)
    targets = _digest_admins(con)
    for t in targets:
        ai_dm_send(con, t, sm["text"])
    c = con.cursor()
    c.execute("INSERT INTO digest_sent(digest_date, count) VALUES(?,?)", (date_str, sm["count"]))
    con.commit()
    audit_log(con, "scheduler", "daily_digest_pushed",
              {"date": date_str, "to": targets, "count": sm["count"]})
    bus.fanout("digest", {"date": date_str, "to": targets, "count": sm["count"]})
    return {"pushed": True, "date": date_str, "to": targets, "count": sm["count"]}

File: im-ai-office/backend-ts/src/digest.ts
import { eq } from "drizzle-orm";
import { db } from "./db/drizzle.js";
import { digestSent, role } from "./db/schema.js";
import { config } from "./config.js";
import { auditLog } from "./repos.js";
import { aiDmSend } from "./aiDm.js";
import { fanout } from "./sse.js";
import { buildDailySummary } from "./memory.js";

// ============ 每日汇总兜底推送（digest.py 的 TS 版；digest_sent 按日期幂等） ============

async function digestAdmins(): Promise<string[]> {
  try {
    const rows = await db.select({ oim_user_id: role.oimUserId })
      .from(role).where(eq(role.role, "admin")).orderBy(role.oimUserId);
    const ids = rows.map((r) => r.oim_user_id).filter(Boolean) as string[];
    return ids.length ? ids : [config.digestFallbackAdmin];
  } catch {
    return [config.digestFallbackAdmin];
  }
}

export async function scanAndPush(now?: Date): Promise<{ pushed: boolean; date: string; reason?: string; to?: string[]; count?: number }> {
  now = now ?? new Date();
  const pad = (x: number) => String(x).padStart(2, "0");
  const dateStr = `${now.getFullYear()}-${pad(now.getMonth() + 1)}-${pad(now.getDate())}`;
  const sent = await db.select({ x: digestSent.digestDate }).from(digestSent)
    .where(eq(digestSent.digestDate, dateStr)).limit(1);
  if (sent.length) {
    return { pushed: false, reason: "already_sent", date: dateStr };
  }
  let gate = new Date(now!);
  const [hh, mm] = config.digestTime.split(":");
  gate.setHours(parseInt(hh, 10) || 18, parseInt(mm, 10) || 0, 0, 0);
  if (now < gate) return { pushed: false, reason: "before_time", date: dateStr };

  const sm = await buildDailySummary(dateStr);
  const targets = await digestAdmins();
  for (const t of targets) await aiDmSend(t, sm.text);
  await db.insert(digestSent).values({ digestDate: dateStr, count: sm.count });
  await auditLog("scheduler", "daily_digest_pushed", { date: dateStr, to: targets, count: sm.count });
  fanout("digest", { date: dateStr, to: targets, count: sm.count });
  return { pushed: true, date: dateStr, to: targets, count: sm.count };
}

